sw/device/sca/sha3_serial.c

```diff
@@ -73,20 +90,286 @@ static void kmac_reset(void) {
 }
 
 /**
- * Issues a process command to KMAC and waits until absorb step is complete.
+ * Report whether the hardware is currently idle.
+ *
+ * If the hardware is not idle then the `CFG` register is locked.
+ *
+ * @param params Hardware parameters.
+ * @returns Whether the hardware is currently idle or not.
  */
-static void kmac_absorb_end(void) {
-  // TODO(#7841, #7842): Remove when we finalize the way we capture traces.
-  uint32_t reg = 0;
-  reg = bitfield_field32_write(reg, KMAC_CMD_CMD_FIELD,
-                               KMAC_CMD_CMD_VALUE_PROCESS);
-  mmio_region_write32(kmac.params.base_addr, KMAC_CMD_REG_OFFSET, reg);
+static bool is_state_idle(dif_kmac_params_t params) {
+  uint32_t reg = mmio_region_read32(params.base_addr, KMAC_STATUS_REG_OFFSET);
+  return bitfield_bit32_read(reg, KMAC_STATUS_SHA3_IDLE_BIT);
+}
+
+/**
+ * Calculate the rate (r) in bits from the given security level.
+ *
+ * @param security_level Security level in bits.
+ * @returns Rate in bits.
+ */
+static uint32_t calculate_rate_bits(uint32_t security_level) {
+  // Formula for the rate in bits is:
+  //
+  //   r = 1600 - c
+  //
+  // Where c is the capacity (the security level in bits multiplied by two).
+  return 1600 - 2 * security_level;
+}
+
+/**
+ * Starts KMAC message without sending START command.
+ *
+ * Based on dif_kmac_mode_kmac_start().
+ *
+ * Unlike dif_kmac_mode_kmac_start(), this function doesn't provide the START
+ * command to the hardware, i.e., just the key is provided and the initial setup
+ * for starting a new message is performed.
+ */
+static dif_kmac_result_t kmac_msg_start(
```

Comment:
nit: I think it would be easier to maintain the code if we modify the kmac dif so that there is a `kmac_mode_kmac_init` that does this, then we add a `start` function separately.  The same comment applies to the other functions below. 

The dif interface may be a little odd, but you can always just keep the current API as is, and then add additional functions there. 

Feel free to open a github issue to handle this as a cleanup task.

Reply:
I wasn't sure if it would be okay to move these functions into the DIF/edit the DIF as the real code should not use these SCA-only functions. But thanks for bringing this up. I agree that this would be better. I've added a new item to the list in #7842 for tracking this.